### ISSUE TITLE
Variable rename

### ISSUE THREAD

**Issue by kyleabeauchamp** [pymc-devs/pymc#2400]
It appears that the symbolic computation graph is ignoring the user-given names given to deterministic variables that happen to evaluate to the same constant.  

```
import pymc3 as pm

with pm.Model():
    a = pm.Uniform("a", 0, 1)
    x = pm.Deterministic("x", pm.math.constant(0.0))
    y = pm.Deterministic("y", pm.math.constant(0.0))
    tr = pm.sample(1000)

print(tr.varnames)

['a_interval__', 'a', 'y', 'y']
```

**Comment by kyleabeauchamp** [comment]
Notice that the "x" variable is deleted from my trace.

**Comment by twiecki** [comment]
Definitely looks like a bug.

**Comment by bwengals** [comment]
~~It looks like `pm.math.constant` is an alias for `theano.tensor.constant`.  From its docstring, the latter is used to~~

>     ... create a small cache of frequently used constant.
>     This speed up the Merge optimization for big graph.
>     We want to cache all scalar to don't merge as frequently constants ...

 ~~I think the right function here is probably `theano.tensor.constant_or_value`.  Maybe `math.py` should provide an alias for this function instead?~~

**Comment by bwengals** [comment]
Disregard my last comment, that's not right.  There have been fixes to `constant` since release 0.0.9.  Maybe `master` doesn't have this problem?  [0.0.9](https://github.com/Theano/Theano/blob/0.9.X/theano/tensor/basic.py#L262) and [master](https://github.com/Theano/Theano/blob/2c6adcccaefb3aa466be76f403bcc4981d023a6a/theano/tensor/basic.py#L208)  I'm on 0.0.9 and get the same error as you.

**Comment by junpenglao** [comment]
Seems this is something to do with theano memory management: `0.0` gets the same memory or something? As the following case works fine:
```python
import pymc3 as pm

with pm.Model() as model:
    a = pm.Uniform("a", 0, 1)
    x = pm.Deterministic("x", pm.math.constant(0.0))
    y = pm.Deterministic("y", pm.math.constant(0)) # <- just changing the number here 
    tr = pm.sample(2)

print(tr.varnames)
```
